### DISCUSSION TITLE
Restart after fork

### DISCUSSION THREAD

**Discussion by GemPG** [erthink/libmdbx#338] [Q&A] [answered]
In "_restrictions.md" you say: 

## Child processes
Do not use opened \ref MDBX_env instance(s) in a child processes after `fork()`.
...
On the other hand, MDBX allow calling \ref mdbx_env_close() in such cases to
release resources, but no more and in general this is a wrong way.

This second sentence is not clear, I think that you mean that in the fork child I MUST NOT call mdbx_env_close() if my parent process has any read transactions.

## Do not open twice
Do not have open an MDBX database twice in the same process at the same
time. 
...
[fallback...] MDBX uses POSIX locks on files, and these
locks have issues if one process opens a file multiple times.

# Question

So if I DO want to reopen the same database in the fork child, do you have a recommended sequence of operations to safely close the old database? "Leaks" are not an issue in fork child.

This sequence must (1) allow the fork child to properly lock the database (2) allow the fork parent to continue any open tables and read operations on other threads to continue safely. I am happy that the forking thread has no transactions, and to in-process mutex any write transaction vs fork. 

Also are the global objects safe to keep using? I.e.:
static mdbx_thread_key_t rthc_key;
static MDBX_atomic_uint32_t rthc_pending;

**Comment by erthink** [comment, marked as answer]
The current version provides the [`mdbx_env_resurrect_after_fork()`](https://libmdbx.dqdkfa.ru/group__c__extra.html#gab7d13c1dbf074bc23ebda2d886add02a) function for this purpose.

---

1. Basically you should completely close database before fork and reopen after, or just don't fork before the open.
Only this way is proper, robust and portable.
2. Alternatively, you may try to close and reopen DB in the child process after the fork.
Such a scenario of work (trick) is not considered a main one, i.e. it is a trick. So I can't guarantee it will work in an all cases.
However, there are several safety checks in the code. Therefore, there is reason to believe that this will work, or it will be possible to gradually eliminate the shortcomings and problems (if will ones).
3. No other options, sure. Unless you will taken into account whole state of all threads working with the particular DB and OS kernel state corresponding to mmapings and used locking scheme (for all thouse processes and threads).
